src/toxcore/net_crypto/mod.rs

```diff
@@ -194,23 +194,48 @@ impl NetCrypto {
     }
 
     /// Add connection to a friend when its DHT `PublicKey` is known.
-    pub fn add_connection(&self, peer_real_pk: PublicKey, peer_dht_pk: PublicKey) {
+    pub fn add_connection(&self, peer_real_pk: PublicKey, peer_dht_pk: PublicKey, saddr: Option<SocketAddr>) {
         let mut connections = self.connections.write();
 
         if connections.contains_key(&peer_real_pk) {
             return;
         }
 
         let dht_precomputed_key = precompute(&peer_dht_pk, &self.dht_sk);
-        let connection = CryptoConnection::new(
+        let mut connection = CryptoConnection::new(
             &dht_precomputed_key,
             self.dht_pk,
             self.real_pk,
             peer_real_pk,
             peer_dht_pk
         );
+        connection.udp_addr = saddr;
```

Comment:
Is there any code where you don't set `connection.udp_addr` right after you have created `CryptoConnection::new`?

Reply:
Yes. It might not be present in Ktree after we found a DHT key. In this case we will find it later. That's why `saddr` is `Option`.